apollo-runtime/src/commonMain/kotlin/com/apollographql/apollo3/network/http/HttpNetworkTransport.kt

```diff
@@ -78,58 +80,90 @@ private constructor(
+        )
       }
+    }
+  }
 
-      emitAll(
-          bodies.map { body ->
-            // On native, body will be freezed in worker.doWork, but reading a BufferedSource is a mutating operation.
-            // So we read the bytes into a ByteString first, and create a new BufferedSource from it after freezing.
-            val bodyByteString = if (platform() == Platform.Native) body.readByteString() else null
-            // Do not capture request
-            val operation = request.operation
-            val response = try {
-              if (bodyByteString != null) {
-                // Native case, use bodyByteString
-                worker.doWork {
-                  operation.parseJsonResponse(
-                      jsonReader = Buffer().apply { write(bodyByteString) }.jsonReader(),
-                      customScalarAdapters = customScalarAdapters
-                  )
-                }
-              } else {
-                // Non-native case, use body
-                worker.doWork {
-                  operation.parseJsonResponse(
-                      jsonReader = body.jsonReader(),
-                      customScalarAdapters = customScalarAdapters
-                  )
-                }
-              }
-            } catch (e: Exception) {
-              throw wrapThrowableIfNeeded(e)
-            }
-
-            @Suppress("DEPRECATION")
-            response.newBuilder()
-                .requestUuid(request.requestUuid)
-                .addExecutionContext(
-                    HttpInfo(
-                        startMillis = millisStart,
-                        endMillis = currentTimeMillis(),
-                        statusCode = httpResponse.statusCode,
-                        headers = httpResponse.headers
-                    )
-                )
-                .build()
-          }
-      )
+  private suspend fun <D : Operation.Data> singleResponse(
+      operation: Operation<D>,
+      customScalarAdapters: CustomScalarAdapters,
+      httpResponse: HttpResponse,
+  ): ApolloResponse<D> {
+    val response = try {
+        worker.doWork {
+          operation.parseJsonResponse(
+              jsonReader = httpResponse.body!!.jsonReader(),
+              customScalarAdapters = customScalarAdapters
+          )
+      }
+    } catch (e: Exception) {
+      throw wrapThrowableIfNeeded(e)
     }
+
+    return response
   }
 
+  @OptIn(ApolloInternal::class)
+  private suspend fun <D : Operation.Data> multipleResponses(
+      operation: Operation<D>,
+      customScalarAdapters: CustomScalarAdapters,
+      httpResponse: HttpResponse,
+  ): Flow<ApolloResponse<D>> {
+    val jsonMerger = DeferredJsonMerger()
+    return multipartBodyFlow(httpResponse).map { part ->
+      try {
+        // On native, components of jsonMerger are frozen inside worker.doWork, making mutability impossible.
+        // So we clone them before they are captured.
```

Comment:
```suggestion
        // On native, we cannot pass `jsonMerger._merge` to `worker.doWork` or it will become frozen making
        // any subsequent payload merge throw.
        // So we clone them before they are captured.
        // XXX: revisit with the new memory model
```

Reply:
Much better! 👍